README.md
# stash

[![Join the chat at https://gitter.im/gosquared/stash](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/gosquared/stash?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

[![Travis](https://api.travis-ci.org/gosquared/stash.svg)](https://travis-ci.org/gosquared/stash)
[![Dependencies](https://david-dm.org/gosquared/stash.svg)](https://david-dm.org/gosquared/stash)
[![Join the chat at https://gitter.im/gosquared/stash](https://img.shields.io/badge/gitter-join%20chat-blue.svg)](https://gitter.im/gosquared/stash)

[![NPM](https://nodei.co/npm/node-stash.png?downloads=true&downloadRank=true&stars=true)](https://www.npmjs.com/package/node-stash)

Distributed cache using Redis as cache store and memory lru for fast access to loaded data.

Requires Redis >= v2.6.12 and `node-redis` client `^0.8`

## Install

    npm install node-stash

## Usage

```javascript

var Stash = require('node-stash');
var Redis = require('redis');
var stash = Stash.createStash(redis.createClient, {
  lru: {
    max: 100000,          // max number of cached results
    maxAge: 600000,       // max age of cached results
    errTTL: 5000,         // max age of cached error results
    timeout: 5000         // min time before callback queue reset
  }
});

var fetchRow = function(id, cb) {
  var fetch = function(done) {
    /**
     * This is our function to fetch data from a resource.
     * The results of this function are cached by stash.
     *
     * It is wise to set a timeout on your fetch, so that
     * it calls `done` with an error if it takes too long
     * to complete. This is useful in case your external
     * resource is overloaded and being slow. If configured,
     * stash will cache this error and prevent overloading the
     * resource with more fetches.
     */
    // Example: querying a mysql db
    mysql.query({
      sql: 'SELECT thing FROM things WHERE id = ? LIMIT 1',
      timeout: 5000
    }, [id], done);
  };

  var key = 'appName:component:'+id;
  stash.get(key, fetch, cb);
};

// get a row
fetchRow(1, function(err, row) {
  // called async with cached or
  // freshly fetched data
});

```

## API

Create an instance of stash:

```javascript
var Stash = require('node-stash');
var stash = Stash.createStash(createRedisClient, { /* options... optional :) */ });
```

`createRedisClient` must be a function that creates and returns an instance of [node-redis](https://github.com/mranney/node_redis) compatible with `v0.8`. This way your app can still have control over instantiating and handling errors on the redis client.

Do not simply return an existing redis client, as this will be placed into pub-sub mode. Stash needs to be able to create distinct redis clients for its pub-sub and non-pub sub Redis functionality.

### stash.get(key, fetchFn, cb)

Retrieve a key. If the key has been previously retrieved, and not older than the LRU maxAge, it will be loaded from memory, resulting in extremely fast access. If not, it will attempt to load the key from Redis. If the key is not in Redis, it will invoke `fetchFn`, which is a function you define to retrieve the value. Stash caches the result of `fetchFn` in Redis.

Stash is designed to run in a distributed architecture. As such, you might have multiple instances of an app running stash, with many processes attempting to retrieve values from the database and cache. Stash has in-built concurrency control around your `fetchFn`, which ensures that only one instance of your app will be able to execute `fetchFn` at a time. This helps prevent stampeding your data source when keys expire from the cache.

If the `fetchFn` invokes its callback with an error, you can configure a different ttl for cached error results. See the `lru.errTTL` option.

`fetchFn` signature: `function (cb) {}`

Example `fetchFn`:

```javascript
var fetch = function(cb) {
  mysql.query('SELECT something FROM somewhere', cb);
};
```

When multiple calls to `stash.get` are made before a key is retrieved and cached, stash will transparently queue callbacks until the first single fetch & cache operation is completed, at which point the callbacks in the queue are invoked with the result. In this way, Stash abstracts concurrency complexity, so you only need to worry about a single function call to retrieve a value for your distributed app.

### stash.del(key, [cb])

Remove a value from the memory and Redis cache. Once this is complete, the value will need to be fetched.

Note that the key is only deleted from the memory cache of the current process, so if you're running many distributed instances of stash, their memory caches might still contain the key. Use `stash.invalidate` to perform cluster-wide purges.

### stash.clear()

Empty the memory cache for all keys. Does not touch Redis cache.

### stash.invalidate(key, [cb])

Invalidate a key from the cache. This is similar to `stash.del`, but goes one step further and ensures that all other instances of stash in other processes also delete the key from their memory caches. An invalidation is a cluster-wide purge.

## Options

```javascript

// example handlers
var debug = console.log;
var metrics = {
  increment: function(name, inc) {
    // name is the counter name, inc is the increment number
  }
};

var opts = {
  redis: {
    wait: true,           // when false, errors if redis connection is down, otherwise queues commands
    ttl: {
      cache: 600000,      // redis cache key ttl
      lock: 10000         // in-case unlock does not work, ttl of redis-based fetchFn lock
    },
    clients: {
      // Clients passed here will be used instead of createRedisClient()
      // This is useful if you want to share redis clients across
      // many instances of stash. It is recommended to create clients
      // solely for use with stash instances.
      cache: null,
      broadcast: null // This client is placed in pub/sub mode.
      // It is recommended to call client.setMaxListeners(0) on this client if
      // you intend to share it across more than 10 instances of stash.
    }
  },
  timeout: {
    retry: 1000           // optimistic lock retry delay
  },
  lru: {
    max: 100000,          // max number of cached results
    maxAge: 600000,       // max age of cached results
    errTTL: 5000,         // max age of cached error results
    timeout: 5000         // min time before callback queue reset
  },
  retryLimit: 5,          // max retry times for optimistic locking
  log: debug,             // set to your own logging handler
  metrics: metrics        // set to your own metrics handler
};
```


CHANGELOG.md
Changelog
---

# v0.4.0

* Upgrade deps.

# v0.3.3

* Remove leaky default metrics handler.

# v0.3.2

* Upgrade [`node-redis-warlock`](https://www.npmjs.org/package/node-redis-warlock) to `^v0.1`.

# v0.3.1

* Replace multi with equivalent SET args.
* Change Redis requirement to `v2.6.12`.
* Options for passing created redis clients for stash to use as an alternative to stash creating its own for each instance.

# v0.3.0

* New lru cache internals.
* `memoize` options changed to `lru`.
* Removed all callback timeouts. Timeouts should be handled by the application when fetching from its resource.

# v0.2.0

* All arguments to `stash.get(key, fetchFn, cb)` are now mandatory.
* Added separate memory cache for errors, allowing for different `ttl` and `max` settings.
* Rearrange options.


test/stash.js
var assert = require('assert');
var Stash = require('../lib/stash');
var Warlock = require('node-redis-warlock');
var async = require('neo-async');
var Redis = require('redis');
var createStash = Stash.createStash;

var mockDbFetch = function(cb) {
  setImmediate(function() {
    cb(null, { test: 1 });
  });
};

var dbErrMsg = 'DB AINT COOL RIGHT NOW';
var mockDbFetchErr = function(cb) {
  setImmediate(function() {
    cb(new Error(dbErrMsg));
  });
};

var createRedisClient = function() {
  return Redis.createClient();
};

describe('Stash', function() {
  var stash = createStash(createRedisClient);
  var redis = stash.cacheRedis;

  describe('get', function() {

    it('errors if no fetch function given', function(done) {
      stash.get('test', null, function(err){
        assert(err);
        done();
      });
    });

    it('no-ops if no cb given', function() {
      stash.get('test');
    });

    it('fetches uncached data', function(done) {
      stash.get('test', mockDbFetch, function(err, data){
        assert(!err);
        assert.equal(1, data.test);
        done();
      });
    });

    it('saved to cache', function(done) {

      redis.get('test', function(err, data) {
        assert(!err);
        assert(data);

        data = stash.decodeContent(data)[0];
        assert.equal(1, data.test);
        done();
      });
    });

    it('fetches cached data', function(done) {
      stash.get('test', function(){
          throw new Error('Fetcher should not be invoked');
        }, function(err, data){
        assert(!err);
        assert(data);

        assert.equal(1, data.test);
        done();
      });
    });
  });

  describe('delete', function() {
    before(function(done) {
      redis.exists('test', function(err, exists) {
        assert(1, exists);

        stash.del('test', done);
      });
    });

    it('should remove from redis', function(done) {
      redis.exists('test', function(err, exists) {
        assert(true, exists);

        done(err);
      });
    });

    it('should clear', function() {
      stash.clear();
    });
  });

  describe('during cache issues', function(done) {
    var stash = createStash(createRedisClient, {
      redis: {
        wait: false
      }
    });

    before(function(done) {
      stash.cacheRedis.end();
      done();
    });

    it('gives error when getting a key', function(done) {
      stash.get('blah', mockDbFetch, function(err, data){
        assert(err);
        assert.equal('redis unavailable', err.message);
        done();
      });
    });

    it('gives error when deleting a key', function(done) {
      stash.del('blah', function(err, data){
        assert(err);
        assert.equal('redis unavailable', err.message);
        done();
      });
    });
  });

  describe('during db issues', function(done) {
    var stash = createStash(createRedisClient);

    it('db error is cached', function(done) {
      var dbFetches = 0;
      var fetch = function(cb) {
        assert.equal(1, (++dbFetches));
        return mockDbFetchErr(cb);
      };

      stash.get('blah1', fetch, function(err1) {
        stash.get('blah1', fetch, function(err2, data){
          assert(!data);
          assert.equal(dbErrMsg, err2.message);
          assert.equal(err2, err1);

          done();
        });
      });
    });
  });

  after(cb => stash.del('blah1', cb));

  describe('concurrency', function() {
    var stash = createStash(createRedisClient);
    var redis = stash.cacheRedis;
    var warlock = Warlock(redis);

    it('sets lock when db fetching', function(done){
      stash.get('sheep', function(cb){
        warlock.lock('sheep', stash.conf.redis.ttl.lock, function(err, unlock){
          assert(!err);

          assert.equal(false, !!unlock);
          cb();
          done();
        });
      }, function() {});
    });

    after(cb => stash.del('sheep', cb));

    it('stops retrying if locked and retry limit reached', function(done) {

      var stash2 = createStash(createRedisClient, {
        retryLimit: 0,
        timeout: {
          retry: 1
        }
      });

      stash.get('retryLimit', function(cb) { }, function(){});

      stash2.get('retryLimit', function(cb){ }, function(err){
        assert.equal('retry limit reached', err.message);
        done();
      });
    });

    after(cb => stash.del('retryLimit', cb));

    it('many gets on uncached key from a single instance result in only one db fetch', function(done) {
      var numGets = 100;
      var fetches = 0;

      var doGet = function(n, next) {
        stash.get('hotKey', function(cb) {
          assert.equal(1, (++fetches));

          return setImmediate(function() {
            cb(null, { test: 2 });
          });
        }, next);
      };

      async.times(numGets, doGet, function(err, results) {
        assert(!err);

        assert.equal(numGets, results.length);
        done();
      });
    });

    after(cb => stash.del('hotKey', cb));

    it('gets from multiple separate instances should only fetch from db once', function(done) {
      var instances = [];
      var numInstances = 5;
      var fetches = 0;

      for (var i = 0; i < numInstances; i++) {
        instances.push(createStash(createRedisClient, {
          timeout: {
            retry: 1
          }
        }));
      }

      async.map(instances, function(stash, done) {
        stash.get('multiInstance', function(cb) {
          assert.equal(1, (++fetches));

          return setImmediate(function() {
            cb(null, { test: 2 });
          });
        }, done);
      }, done);
    });
  });

  after(cb => stash.del('multiInstance', cb));

  describe('broadcast', function() {
    var stash = createStash(createRedisClient);
    var redis = stash.cacheRedis;

    var stash2 = createStash(createRedisClient);

    before(function(done) {
      // precache
      stash.get('pizza', mockDbFetch, function(err, data){
        assert(!err);
        assert(data);

        assert.equal(1, data.test);

        // Make sure cache key is actually there
        redis.get('pizza', function(err, data){
          assert(!err);
          assert(data);

          stash2.get('pizza', mockDbFetch, function(err, data) {
            assert.equal(1, data.test);
            return done(err);
          });
        });
      });
    });

    it('should invalidate local cache for all instances', function(done) {
      var numFetches = 0;

      var fetch = function(cb) {
        numFetches += 1;
        setImmediate(cb);
      };

      stash2.broadcast.once('message', function(){
        var purgeTest = function() {
          stash2.get('pizza', fetch, function() {
            assert.equal(1, numFetches);
            return done();
          });
        };

        setImmediate(purgeTest);
      });

      stash.invalidate('pizza');
    });

    after(cb => stash.del('pizza', cb));

  });
});


test/setup/redisFlush.js

before(function(done){
  this.redis.flushall(done);
});


lib/stash.js
var extend        = require('extend');
var Broadcast     = require('./broadcast');
var Warlock       = require('node-redis-warlock');
var Rush          = require('node-rush');
var debug         = require('debug')('stash');

var metrics = { increment: function() {} };
var slice = Array.prototype.slice;

var Stash = function(createRedisClient, opts) {
  var self = this;

  var conf = {
    redis: {
      wait: true,           // when false, errors if redis connection is down, otherwise queues commands
      ttl: {
        cache: 600000,      // redis cache key ttl
        lock: 10000         // in-case unlock does not work, ttl of redis-based fetchFn lock
      },
      clients: {
        // Clients passed here will be used instead of createRedisClient()
        // This is useful if you want to share redis clients across
        // many instances of stash. It is recommended to create clients
        // solely for use with stash instances.
        cache: null,
        broadcast: null // This client is placed in pub/sub mode.
        // It is recommended to call client.setMaxListeners(0) on this client if
        // you intend to share it across more than 10 instances of stash.
      }
    },
    timeout: {
      retry: 1000          // optimistic lock retry delay
    },
    lru: {
      max: 100000,          // max number of cached results
      maxAge: 600000,       // max age of cached results
      errTTL: 5000,         // max age of cached error results
      timeout: 5000         // min time before callback queue reset
    },
    retryLimit: 5,          // max retry times for optimistic locking
    log: debug,             // set to your own logging handler
    metrics: metrics        // set to your own metrics handler
  };

  self.conf = extend(true, conf, opts || {});
  self.log = conf.log;
  self.metrics = conf.metrics;
  self.cacheRedis = self.conf.redis.clients.cache || createRedisClient();
  self.broadcastRedis = self.conf.redis.clients.broadcast || createRedisClient();
  self.warlock = Warlock(self.cacheRedis);
  self.lru = Rush(self.conf.lru);

  self.broadcast = Broadcast(self, {
    redis: self.cacheRedis,
    subscriberRedis: self.broadcastRedis,
    log: self.log,
    metrics: self.metrics
  });
};

Stash.prototype.get = function(key, dbFetch, cb) {
  var self = this;

  if (typeof cb !== 'function') {
    // no callback given so not sure exactly what arguments have been
    // supplied. Safest to no-op than to assume dbFetch is actually
    // a fetch function.
    return self.log('stash.get requires callback');
  }

  if (typeof dbFetch !== 'function') {
    var err = new Error('no fetch function given');
    setImmediate(function() {
      cb(err);
    });
    return;
  }

  self.metrics.increment('get');

  function cacheFetch(cb) {
    self.cacheFetch(key, dbFetch, 0, cb);
  }

  self.lru.get(key, cacheFetch, cb);
};

/** Neither result nor error cached in memory, so fetch from cache */
Stash.prototype.cacheFetch = function(key, dbFetch, retries, cb) {
  var self = this;

  var done = function fetchComplete(err, result){
    cb.apply(null, [err].concat(result));
  };

  self.metrics.increment('fetch');

  if (!self.cacheRedis.connected && !self.conf.redis.wait) {
    return setImmediate(function() {
      done(new Error('redis unavailable'));
    });
  }

  self.cacheRedis.get(key, function(err, content) {
    if (err) return done(self.reportCacheErr(err));
    var result;

    if (content) {
      self.log('cache:hit');
      self.metrics.increment('hit');
      try {
        result = self.decodeContent(content);
      } catch (e) {
        self.log('cache:decode exception');
        err = e;
      }

      return done(err, result);
    }

    self.log('cache:miss');
    self.metrics.increment('miss');

    var lock = function(done) {
      self.log('cache:locking');
      self.warlock.lock(
        key,
        self.conf.redis.ttl.lock,
        done
      );
    };

    // Before invoking the db fetch, set a lock
    lock(function(err, unlock) {
      if (err) return done(self.reportCacheErr(err));

      if (typeof unlock !== 'function') {
        // No unlock function given so we didn't obtain a lock
        self.log('cache:locked');

        if (retries++ >= self.conf.retryLimit) {
          self.log('cache:retry limit reached');
          return done(new Error('retry limit reached'));
        }
        self.log('cache:retry');

        // Try fetching from redis cache again after a short pause
        // Hopefully the data will be cached by the time we retry
        setTimeout(function() {
          self.cacheFetch(key, dbFetch, retries, cb);
        }, self.conf.timeout.retry);

        return;
      }

      // Lock in place. We are now responsible for fetching
      // from DB and saving to the cache
      self.log('cache:lockAcquired');

      self.log('db:fetch');
      return dbFetch(function(err) {
        var content;
        var result = slice.call(arguments, 1);

        if (err) {
          // Don't save err in redis but will still save to memory
          self.log('db:error');
          debug(err);
          return unlock(function() {
            done(err, result);
          });
        }

        // Store result args as array in cache
        try {
          content = self.encodeContent(result);
        } catch (e) {
          self.log('cache:encode exception');
          err = new Error('encode exception');
          err.exception = e;
          return unlock(function() {
            // only cache results, not encode error
            done(null, result);
          });
        }

        self.cacheRedis.set(
          key, content,
          'PX', self.conf.redis.ttl.cache,
          function(err) {
            if (err) {
              self.log('cache:save failed');
              debug(err);
            } else {
              self.log('cache:saved');
            }

            self.metrics.increment('stored');
            unlock(function() {
              // cache result even if failed to save to cache
              done(null, result);
            });
          }
        );
      });
    });
  });
};

Stash.prototype.reportCacheErr = function(err) {
  var self = this;
  self.log('cache:error');

  if (err.message === 'timeout') {
    self.metrics.increment('cache.timeout');
    self.log('cache:timeout');
  }

  self.metrics.increment('cache.error');
  return err;
};

Stash.prototype.del = function(key, cb) {
  var self = this;
  cb = cb || function() {};
  self.log('cache:del');
  self.metrics.increment('del');

  if (!self.cacheRedis.connected && !self.conf.redis.wait) {
    self.log('cache:del failed');
    return setImmediate(function() {
      cb(new Error('redis unavailable'));
    });
  }

  self.cacheDel(key, function(err) {
    if (err) return cb(err);

    self.memDel(key);
    cb();
  });
};

Stash.prototype.memDel = function(key) {
  var self = this;
  self.lru.del(key);
};

Stash.prototype.cacheDel = function(key, cb) {
  var self = this;
  self.cacheRedis.del(key, function(err) {

    if (err) {
      self.log('cache:del failed');
      debug(err);
      return cb(err);
    } else {
      self.log('cache:del success');
      self.metrics.increment('deleted');
      return cb(err);
    }
  });
};

Stash.prototype.invalidate = function(key, cb) {
  var self = this;
  cb = cb || function() {};

  self.metrics.increment('invalidate');
  self.del(key, function(err) {
    if (err) return cb(err);

    self.broadcast.publish('invalidateStash', {
      key: key
    }, cb);
  });
};

Stash.prototype.clear = function() {
  var self = this;

  self.metrics.increment('clear');
  self.lru.reset();
};

// Override if required
Stash.prototype.encodeContent = JSON.stringify;
Stash.prototype.decodeContent = JSON.parse;

var createStash = function(createRedisClient, opts) {
  return new Stash(createRedisClient, opts);
};

exports.Stash = Stash;
exports.createStash = createStash;
